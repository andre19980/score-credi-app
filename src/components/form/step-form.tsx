import { useState } from "react";
import { SubmitHandler } from "react-hook-form";

import ProgressLine from "./step-progress";
import Form from "./form";
import PersonalInfoStep from "./steps/personal-info";
import AddressStep from "./steps/address";
import IncomeStep from "./steps/income";

import { FormType } from "./types";

export enum STEPS {
  PERSONAL = 0,
  ADDRESS = 1,
  INCOME = 2,
}

export const FORM_STEPS = [
  {
    id: STEPS.PERSONAL,
    fields: ["name", "age"],
    component: <PersonalInfoStep />,
  },
  {
    id: STEPS.ADDRESS,
    fields: ["city"],
    component: <AddressStep />,
  },
  { id: STEPS.INCOME, fields: ["income"], component: <IncomeStep /> },
];

export default function StepForm() {
  const [currentStep, updateStep] = useState(0);

  const handleNextStep = async (e: React.MouseEvent<HTMLElement>, canGoNext: boolean) => {
    e.preventDefault();

    if (!canGoNext) return;

    if (currentStep < FORM_STEPS.length - 1) {
      updateStep((curStep) => curStep + 1);
    }
  };

  const handlePrevStep = () => {
    if (currentStep > 0) {
      updateStep((curStep) => curStep - 1);
    }
  };

  return (
    <div className="flex flex-col justify-center items-center gap-y-16 sm:col-start-2 sm:col-span-2">
      <ProgressLine currentStep={currentStep} />
      <div className="bg-gray-50 w-full p-6 rounded">
        <Form
          currentStep={currentStep}
          updateStep={updateStep}
          handleNextStep={handleNextStep}
          handlePrevStep={handlePrevStep}
        />
      </div>
    </div>
  );
}
