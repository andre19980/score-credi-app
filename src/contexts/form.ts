import { z } from "zod";
import { createContext } from "react";
import {
  type UseFormHandleSubmit,
  type UseFormRegister,
  type UseFormReset,
  type FieldErrors,
  type SubmitHandler,
  type UseFormTrigger,
} from "react-hook-form";

import { FormDataSchema } from "@/lib/schema";

export type FormType = z.infer<typeof FormDataSchema>;

export interface StepFormContextType {
  currentStep: number;
  updateStep: React.Dispatch<React.SetStateAction<number>>;
  handleNextStep: (e: React.MouseEvent<HTMLElement>, canGoNext: boolean) => Promise<void>;
  handlePrevStep: () => void;
  register: UseFormRegister<FormType>;
  trigger: UseFormTrigger<FormType>;
  handleSubmit: UseFormHandleSubmit<FormType>;
  onSubmit: SubmitHandler<FormType>;
  reset: UseFormReset<FormType>;
  errors: FieldErrors<FormType>;
}

export const StepFormContext = createContext<StepFormContextType | null>(null);
