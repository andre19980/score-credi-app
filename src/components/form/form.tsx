import { createContext, useContext } from "react";
import { zodResolver } from "@hookform/resolvers/zod";
import { useForm, SubmitHandler } from "react-hook-form";

import Actions from "./actions";
import { FORM_STEPS } from "./step-form";

import { FormDataSchema } from "@/lib/schema";
import { UserContext } from "@/app/page";
import { StepFormContextType, FormType } from "./types";

export const StepFormContext = createContext<StepFormContextType | null>(null);

export interface FormComponentPropsType extends Pick<StepFormContextType, "currentStep" | "updateStep" | "handleNextStep" | "handlePrevStep"> {}

export default function Form({
  currentStep,
  updateStep,
  handleNextStep,
  handlePrevStep
}: FormComponentPropsType) {
  const userContext = useContext(UserContext);
  const {
    register,
    handleSubmit,
    reset,
    trigger,
    formState: { errors },
  } = useForm<FormType>({
    resolver: zodResolver(FormDataSchema),
  });

  const onSubmit: SubmitHandler<FormType> = async (formData) => {
    const res = await fetch("http://localhost:8080/api/credito-analise", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(formData),
    });

    const data = await res.json();

    if (data.status === "APPROVED") {
      const val = `true;${data.max_amount.toString()}`;

      sessionStorage.setItem("user_approved", val);
      userContext?.setUser(val);
    } else {
      const val = "false";

      sessionStorage.setItem("user_approved", val);
      userContext?.setUser(val);
    }
  };
  
  const formContext = {
    currentStep,
    updateStep,
    handleNextStep,
    handlePrevStep,
    register,
    handleSubmit,
    onSubmit,
    reset,
    trigger,
    errors,
  } as StepFormContextType;

  return (
    <StepFormContext.Provider value={formContext}>
      <form onSubmit={handleSubmit(onSubmit)}>
        {typeof currentStep === "number" &&
          FORM_STEPS[currentStep].component}
        <Actions />
      </form>
    </StepFormContext.Provider>
  );
}
