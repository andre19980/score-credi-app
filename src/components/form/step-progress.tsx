import { STEPS } from "./step-form";
import { PersonalSVG, AddressSVG, IncomeSVG } from "./icons";

export default function ProgressLine({ currentStep }: { currentStep: number }) {
  const icons = {
    [STEPS.PERSONAL]: {
      svg: <PersonalSVG />,
    },
    [STEPS.ADDRESS]: {
      svg: <AddressSVG />,
    },
    [STEPS.INCOME]: {
      svg: <IncomeSVG />,
    },
  };

  return (
    <ol className="flex items-center w-full px-4 sm:px-16">
      <IconStep
        icon={icons[STEPS.PERSONAL].svg}
        isComplete={currentStep >= 0}
      />
      <IconStep icon={icons[STEPS.ADDRESS].svg} isComplete={currentStep >= 1} />
      <IconStep icon={icons[STEPS.INCOME].svg} isComplete={currentStep >= 2} />
    </ol>
  );
}

interface IconStepType {
  icon: JSX.Element;
  isComplete: boolean;
}

function IconStep({ icon, isComplete }: IconStepType) {
  const incomplete =
    "bg-gray-100 dark:bg-gray-700 text-gray-500 dark:text-gray-100";
  const completed =
    "bg-blue-100 dark:bg-blue-800 text-blue-600 dark:text-blue-500";

  const after =
    "after:content-[''] after:w-full after:h-1 after:border-b after:border-4 after:inline-block";
  const before =
    "before:content-[''] before:w-full before:h-1 before:border-b before:border-4 before:inline-block";

  return (
    <li
      className={`${before} first:before:content-none last:after:content-none flex w-full items-center ${after}`}
    >
      <span
        className={`${isComplete ? completed : incomplete} flex items-center justify-center w-10 h-10 rounded-full lg:h-12 lg:w-12 shrink-0`}
      >
        {icon}
      </span>
    </li>
  );
}
