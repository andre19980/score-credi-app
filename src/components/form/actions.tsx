import { useContext } from "react";

import { FORM_STEPS } from "./step-form";
import { StepFormContext, FormType } from "@/contexts/form";

export default function Actions() {
  const context = useContext(StepFormContext);
  const currentStep = context?.currentStep

  if (currentStep === undefined) return null;

  const isLastStep =  currentStep === FORM_STEPS.length - 1;
  const isFirstStep = currentStep === 0;

  const handleClick = async (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
    type FieldName = keyof FormType;

    const fields = FORM_STEPS[currentStep].fields;
    const output = await context?.trigger(fields as FieldName[], { shouldFocus: true });
  
    context?.handleNextStep(e, !!output);
  }

  return (
    <div className="flex justify-between pt-4 mt-8 border-t">
      <button
        type="button"
        className={`${isFirstStep ? "text-gray-600 bg-gray-100 hover:text-gray-600" : "text-gray-900 bg-white hover:text-blue-700"} py-2.5 px-5 text-sm font-medium focus:outline-none rounded-lg border border-gray-200 hover:bg-gray-100 focus:z-10 focus:ring-4 focus:ring-gray-100 dark:focus:ring-gray-700 dark:bg-gray-800 dark:text-gray-400 dark:border-gray-600 dark:hover:text-white dark:hover:bg-gray-700`}
        onClick={context?.handlePrevStep}
        disabled={isFirstStep}
      >
        Voltar
      </button>

      {isLastStep ? (
        <button
          type="submit"
          className="text-white bg-blue-700 hover:bg-blue-800 focus:ring-4 focus:ring-blue-300 font-medium rounded-lg text-sm px-5 py-2.5 dark:bg-blue-600 dark:hover:bg-blue-700 focus:outline-none dark:focus:ring-blue-800"
        >
          Registrar
        </button>
      ) : (
        <button
          type="button"
          className="text-white bg-blue-700 hover:bg-blue-800 focus:ring-4 focus:ring-blue-300 font-medium rounded-lg text-sm px-5 py-2.5 dark:bg-blue-600 dark:hover:bg-blue-700 focus:outline-none dark:focus:ring-blue-800"
          onClick={handleClick}
        >
          Pr√≥ximo
        </button>
      )}
    </div>
  );
}
