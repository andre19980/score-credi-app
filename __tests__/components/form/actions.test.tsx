import "@testing-library/jest-dom";
import { render } from "@testing-library/react";

import Actions from "@/components/form/actions";
import { StepFormContext } from "@/components/form/form";
import { StepFormContextType } from "@/components/form/types";

const customRender = (ui: JSX.Element, { providerProps, ...renderOptions }: { providerProps: StepFormContextType }) => {
  return render(
    <StepFormContext.Provider value={providerProps}>{ui}</StepFormContext.Provider>,
    renderOptions,
  )
}


describe("Actions Form Component", () => {
  const defaultProps = {
    currentStep: 0,
    updateStep: jest.fn(),
    handleNextStep: jest.fn(),
    handlePrevStep: jest.fn(),
    register: jest.fn(),
    trigger: jest.fn(),
    handleSubmit: jest.fn(),
    onSubmit: jest.fn(),
    reset: jest.fn(),
    errors: {},
  }

  describe("render steps components", () => {
    let providerProps: StepFormContextType;

    it("Step Personal Info", () => {
      const { container } = customRender(<Actions />, { providerProps: defaultProps });
      expect(container).toMatchSnapshot();
    });
  
    it("Step Address case", () => {
      providerProps = {
        ...defaultProps,
        currentStep: 1,
      };

      const { container } = customRender(<Actions />, { providerProps });
      expect(container).toMatchSnapshot();
    });
  
    it("Step Income", () => {
      providerProps = {
        ...defaultProps,
        currentStep: 2,
      };

      const { container } = customRender(<Actions />, { providerProps });
      expect(container).toMatchSnapshot();
    });
  })
});
