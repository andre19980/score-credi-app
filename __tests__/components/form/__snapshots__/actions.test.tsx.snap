// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Actions Form Component render steps components Step Address case 1`] = `
<div>
  <div
    class="flex justify-between pt-4 mt-8 border-t"
  >
    <button
      class="text-gray-900 bg-white py-2.5 px-5 text-sm font-medium focus:outline-none rounded-lg border border-gray-200 hover:bg-gray-100 hover:text-blue-700 focus:z-10 focus:ring-4 focus:ring-gray-100 dark:focus:ring-gray-700 dark:bg-gray-800 dark:text-gray-400 dark:border-gray-600 dark:hover:text-white dark:hover:bg-gray-700"
      type="button"
    >
      Voltar
    </button>
    <button
      class="text-white bg-blue-700 hover:bg-blue-800 focus:ring-4 focus:ring-blue-300 font-medium rounded-lg text-sm px-5 py-2.5 dark:bg-blue-600 dark:hover:bg-blue-700 focus:outline-none dark:focus:ring-blue-800"
      type="button"
    >
      Próximo
    </button>
  </div>
</div>
`;

exports[`Actions Form Component render steps components Step Income 1`] = `
<div>
  <div
    class="flex justify-between pt-4 mt-8 border-t"
  >
    <button
      class="text-gray-900 bg-white py-2.5 px-5 text-sm font-medium focus:outline-none rounded-lg border border-gray-200 hover:bg-gray-100 hover:text-blue-700 focus:z-10 focus:ring-4 focus:ring-gray-100 dark:focus:ring-gray-700 dark:bg-gray-800 dark:text-gray-400 dark:border-gray-600 dark:hover:text-white dark:hover:bg-gray-700"
      type="button"
    >
      Voltar
    </button>
    <button
      class="text-white bg-blue-700 hover:bg-blue-800 focus:ring-4 focus:ring-blue-300 font-medium rounded-lg text-sm px-5 py-2.5 dark:bg-blue-600 dark:hover:bg-blue-700 focus:outline-none dark:focus:ring-blue-800"
      type="submit"
    >
      Registrar
    </button>
  </div>
</div>
`;

exports[`Actions Form Component render steps components Step Personal Info 1`] = `
<div>
  <div
    class="flex justify-between pt-4 mt-8 border-t"
  >
    <button
      class="text-gray-600 bg-gray-100 py-2.5 px-5 text-sm font-medium focus:outline-none rounded-lg border border-gray-200 hover:bg-gray-100 hover:text-blue-700 focus:z-10 focus:ring-4 focus:ring-gray-100 dark:focus:ring-gray-700 dark:bg-gray-800 dark:text-gray-400 dark:border-gray-600 dark:hover:text-white dark:hover:bg-gray-700"
      disabled=""
      type="button"
    >
      Voltar
    </button>
    <button
      class="text-white bg-blue-700 hover:bg-blue-800 focus:ring-4 focus:ring-blue-300 font-medium rounded-lg text-sm px-5 py-2.5 dark:bg-blue-600 dark:hover:bg-blue-700 focus:outline-none dark:focus:ring-blue-800"
      type="button"
    >
      Próximo
    </button>
  </div>
</div>
`;
